### PR TITLE
feat(GCS+gRPC): implement ListObjects()

### LINKED ISSUE #4197: Implement storage::GrpcClient::ListObject
Remember that this must support prefixes, startOffset, endOffset, delimiter, etc.

### PR DESCRIPTION
**Pull request by coryan**
Fixes #4197

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8056)

### CODE REVIEW

**Comment by coryan** [comment]
Ping, in case folks missed this one.

File: google/cloud/storage/internal/grpc_object_request_parser.cc [hunk 1]
```diff
@@ -260,6 +260,44 @@ ResumableUploadResponse GrpcObjectRequestParser::FromProto(
   return response;
 }
 
+google::storage::v2::ListObjectsRequest GrpcObjectRequestParser::ToProto(
+    ListObjectsRequest const& request) {
+  google::storage::v2::ListObjectsRequest result;
+  result.set_parent("projects/_/buckets/" + request.bucket_name());
+  auto const page_size = request.GetOption<MaxResults>().value_or(0);
+  // Clamp out of range values, the service will clamp to its own range ([0,
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc_object_request_parser.cc:268]: nit: s/, the/. The/

I think the clamping of `page_size` might be overly cautious.... but I trust your judgment. If you want to be super-duper cautious you could add tests for the different cases. (I am not recommending this).

File: google/cloud/storage/internal/grpc_object_request_parser_test.cc [hunk 2]
```diff
@@ -305,6 +306,58 @@ TEST(GrpcObjectRequestParser, InsertObjectMediaRequestWithObjectMetadata) {
...
+      )pb",
+      &expected));
+
+  ListObjectsRequest req("test-bucket");
+  req.set_page_token("test-only-invalid");
+  req.set_multiple_options(
+      MaxResults(10), Delimiter("/"), IncludeTrailingDelimiter(true),
+      Prefix("test/prefix"), Versions(true), StartOffset("test/prefix/a"),
+      EndOffset("test/prefix/abc"), UserProject("test-user-project"),
+      QuotaUser("test-quota-user"), UserIp("test-user-ip"));
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc_object_request_parser_test.cc:332]: `QuotaUser` and `UserIp` are not used... but I see them unused in other tests, so there is probably a reason they are here.

**Review by dbolduc — APPROVED** [review]

File: google/cloud/storage/internal/grpc_object_request_parser_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_object_request_parser_test.cc:332]: Ack, yes, they are intentional (mostly checking for "no crashes and no unexpected changes due to these extra options".

File: google/cloud/storage/internal/grpc_object_request_parser.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_object_request_parser.cc]: Ack.

**Pull request merged**